Changes to support Raw IP packets

cameron-elliott opened the pull request:
Raw IP packets are IPv4 and IPv6 directly inside the raw libpcap packet, the first
byte of the Raw IP packet is the first byte of the IPv4 or IPv6 packet.

This link provides a little information: http://www.tcpdump.org/linktypes.html
[Search for LINKTYPE_RAW]

This is the file format you get when you use Rawcap.exe to capture packets on the loopback interface.
More: http://www.netresec.com/?page=RawCap
[Winpcap cannot normally capture off of the loopback interface, so this is a nice tool to capture loopback packets, and with the added support for Raw linklayer, you can now parse these with Packetnet]

It is quite possible this is the link layer used by Wireshark and other tools on Linux when capturing from the loopback interface, although they can use LinuxSLL as an option.

chmorgan commented:
This looks like a pretty good start. I cleaned some things up and pushed the new branch out to: https://github.com/chmorgan/packetnet/tree/cameron_pr

A few things:

- Its helpful to have a unit test for each type of packet/datagram being parsed. Can you add one? I'd prefer all new parsers come with a unit test so we don't regress when refactoring other code.

- I modified the constructor to clean it up a bit, can you verify that it looks like you want? Effectively because there is no header used I dropped the Parsexxx() method.

- The >>4 in the constructor isn't immediately clear. Its typically better to use a defined number so there is some documentation in the code. If you aren't clear I can try to clean that up as I think I understand why you are doing that.

I also removed some extra whitespace and fixed line endings in the .csproj file.

cameron-elliott commented:
Awesome, thanks for the suggestions, I will work on these and submit
another PR.
Cameron

Cameron Elliott
Lime Video Software  206-310-3284
www.limevideo.com
Digital Video SDKs for C# - .NET for Windows - Mono for Linux

On Wed, Nov 30, 2016 at 10:35 AM, chmorgan <notifications@github.com> wrote:

> This looks like a pretty good start. I cleaned some things up and pushed
> the new branch out to: https://github.com/chmorgan/
> packetnet/tree/cameron_pr
>
> A few things:
>
>    -
>
>    Its helpful to have a unit test for each type of packet/datagram being
>    parsed. Can you add one? I'd prefer all new parsers come with a unit test
>    so we don't regress when refactoring other code.
>    -
>
>    I modified the constructor to clean it up a bit, can you verify that
>    it looks like you want? Effectively because there is no header used I
>    dropped the Parsexxx() method.
>    -
>
>    The >>4 in the constructor isn't immediately clear. Its typically
>    better to use a defined number so there is some documentation in the code.
>    If you aren't clear I can try to clean that up as I think I understand why
>    you are doing that.
>
> I also removed some extra whitespace and fixed line endings in the .csproj
> file.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/chmorgan/packetnet/pull/12#issuecomment-263955629>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA1BUS3DI55Y2rHhwIhuNaAi30vBGQWlks5rDcH4gaJpZM4K_mp0>
> .
>

chmorgan commented:
Please base it on the branch I posted there, cameron_pr.

Chris

On Wed, Nov 30, 2016 at 3:34 PM Cameron Elliott <notifications@github.com>
wrote:

> Awesome, thanks for the suggestions, I will work on these and submit
> another PR.
> Cameron
>
> Cameron Elliott
> Lime Video Software 206-310-3284
> www.limevideo.com
> Digital Video SDKs for C# - .NET for Windows - Mono for Linux
>
> On Wed, Nov 30, 2016 at 10:35 AM, chmorgan <notifications@github.com>
> wrote:
>
> > This looks like a pretty good start. I cleaned some things up and pushed
> > the new branch out to: https://github.com/chmorgan/
> > packetnet/tree/cameron_pr
> >
> > A few things:
> >
> > -
> >
> > Its helpful to have a unit test for each type of packet/datagram being
> > parsed. Can you add one? I'd prefer all new parsers come with a unit test
> > so we don't regress when refactoring other code.
> > -
> >
> > I modified the constructor to clean it up a bit, can you verify that
> > it looks like you want? Effectively because there is no header used I
> > dropped the Parsexxx() method.
> > -
> >
> > The >>4 in the constructor isn't immediately clear. Its typically
> > better to use a defined number so there is some documentation in the
> code.
> > If you aren't clear I can try to clean that up as I think I understand
> why
> > you are doing that.
> >
> > I also removed some extra whitespace and fixed line endings in the
> .csproj
> > file.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/chmorgan/packetnet/pull/12#issuecomment-263955629>,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AA1BUS3DI55Y2rHhwIhuNaAi30vBGQWlks5rDcH4gaJpZM4K_mp0
> >
> > .
> >
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/chmorgan/packetnet/pull/12#issuecomment-263987338>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABJ-APCupQMdcYX9rVOJH4zSPn5kXEUwks5rDd3qgaJpZM4K_mp0>
> .
>

cameron-elliott commented:
Chris,

I hope you are well.

I have posted an updated PR.

Okay, I believe I have responded to your points in the new PR, but
basically:
- I added test cases
- I like the new constructor too, and I fixed an issue also
- I added some comments to what the >>4 is all about
- I have run all the test in VS2015 and in appveyor, and they all pass.
https://goo.gl/ekTyrh



I did have an appveyor.yml, but I removed it from this PR, I think I will
pitch that and building nuget packages in a later PR.

Please let me know your thoughts.
Cameron



Cameron Elliott
Lime Video Software  206-310-3284
www.limevideo.com
Digital Video SDKs for C# - .NET for Windows - Mono for Linux

On Wed, Nov 30, 2016 at 10:35 AM, chmorgan <notifications@github.com> wrote:

> This looks like a pretty good start. I cleaned some things up and pushed
> the new branch out to: https://github.com/chmorgan/
> packetnet/tree/cameron_pr
>
> A few things:
>
>    -
>
>    Its helpful to have a unit test for each type of packet/datagram being
>    parsed. Can you add one? I'd prefer all new parsers come with a unit test
>    so we don't regress when refactoring other code.
>    -
>
>    I modified the constructor to clean it up a bit, can you verify that
>    it looks like you want? Effectively because there is no header used I
>    dropped the Parsexxx() method.
>    -
>
>    The >>4 in the constructor isn't immediately clear. Its typically
>    better to use a defined number so there is some documentation in the code.
>    If you aren't clear I can try to clean that up as I think I understand why
>    you are doing that.
>
> I also removed some extra whitespace and fixed line endings in the .csproj
> file.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/chmorgan/packetnet/pull/12#issuecomment-263955629>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA1BUS3DI55Y2rHhwIhuNaAi30vBGQWlks5rDcH4gaJpZM4K_mp0>
> .
>

chmorgan commented:
Hi Cameron. I'm going to close this PR as it looks like these changes were merged in. I'm still interested in the appveyor changes if you get back to them.
